File: benchmarks/quicksort.py
import sys
sys.setrecursionlimit(20000)

def qsort(lst, lo, hi):
    if lo >= hi:
        return
    pivot = lst[hi]
    i = lo
    for j in range(lo, hi):
        if lst[j] <= pivot:
            lst[i], lst[j] = lst[j], lst[i]
            i += 1
    lst[i], lst[hi] = lst[hi], lst[i]
    qsort(lst, lo, i - 1)
    qsort(lst, i + 1, hi)

lst = []
x = 12345
for i in range(10000):
    x = (x * 1103515245 + 12345) % 2147483647
    lst.append(x % 10000)
qsort(lst, 0, len(lst) - 1)
print(lst[5000])

File: benchmarks/cpp/quicksort.cpp
#include <iostream>
#include <vector>
void qsort_impl(std::vector<long long>& lst, int lo, int hi) {
    if (lo >= hi) return;
    long long pivot = lst[hi];
    int i = lo;
    for (int j = lo; j < hi; j++) {
        if (lst[j] <= pivot) { std::swap(lst[i], lst[j]); i++; }
    }
    std::swap(lst[i], lst[hi]);
    qsort_impl(lst, lo, i - 1);
    qsort_impl(lst, i + 1, hi);
}
int main() {
    std::vector<long long> lst;
    long long x = 12345;
    for (int i = 0; i < 10000; i++) {
        x = (x * 1103515245LL + 12345) % 2147483647LL;
        lst.push_back(x % 10000);
    }
    qsort_impl(lst, 0, (int)lst.size() - 1);
    std::cout << lst[5000] << "\n";
}
